### DISCUSSION TITLE
Docs URL fails with pydatitc 2.4.2

### DISCUSSION THREAD

**Discussion by atompie** [fastapi/fastapi#10460] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Optional, Dict
from fastapi import FastAPI
from pydantic import BaseModel
from pydantic import ConfigDict


class FlowResponse(Dict[str, dict]):

    def __init__(self, *args, **kwargs):
        dict.__init__(self, *args, **kwargs)

    def __setitem__(self, key, value: dict):
        raise ValueError("FlowResponse only accepts dict objects as values.")


class FlowGraph(BaseModel):
    id: Optional[str] = None
    response: Optional[FlowResponse] = FlowResponse()

    model_config = ConfigDict(arbitrary_types_allowed=True)


app = FastAPI()


@app.get("/")
async def root(flow: FlowGraph):
    return {"message": "Hello World"}


if __name__ == "__main__":
    import uvicorn

    uvicorn.run("app.main:app", host="0.0.0.0", port=8181, log_level='info', workers=1)
```

### Description

I'm currently facing an issue that affects both documentation and Pydantic models, and I'm uncertain which part of the project is responsible for it.

Upon upgrading from Pydantic version 2.3.0 to 2.4.0, I observed that endpoints using `arbitrary_types` in the endpoints were causing documentation errors and not rendering `docs` correctly. This issue is particularly noticeable in the `BaseModel` example provided below:

```python
class FlowGraph(BaseModel):
    description: Optional[str] = None
    response: Optional[FlowResponse] = FlowResponse()
    model_config = ConfigDict(arbitrary_types_allowed=True)
```

The `FlowResponse` class is defined as follows:

```python
class FlowResponse(Dict[str, dict]):

    def __init__(self, *args, **kwargs):
        dict.__init__(self, *args, **kwargs)

    def __setitem__(self, key, value: dict):
        if not isinstance value, dict:
            raise ValueError("FlowResponse only accepts dict objects as values.")

        dict.__setitem__(self, key, value)
```

When I use this as a parameter in any endpoint, I encounter a red alert indicating that `openapi.json` cannot be created.

Here are the versions of the installed packages:

- `pydantic` 2.4.2
- `pydantic_core` 2.10.1
- `pydantic-extra-types` 2.1.0
- `pydantic-settings` 2.0.3
- `fastapi` 0.103.2"

Run the code above and enter http://api-url:8181/docs

you should see: Internal Server Error /openapi.json

### Operating System

Linux

### Operating System Details

- `pydantic` 2.4.2
- `pydantic_core` 2.10.1
- `pydantic-extra-types` 2.1.0
- `pydantic-settings` 2.0.3
- `fastapi` 0.103.2"
- python 3.10

### FastAPI Version

0.103.2

### Pydantic Version

2.4.2

### Python Version

3.10

**Comment by Ale-Cas** [comment]
Hi @atompie, 
I was able to replicate your issue but I have some questions: 
1. what's your use case? 
2. why are you trying to pass a required body as a parameter of a get request?
I'm asking because maybe there's another implementation that wouldn't require you to come across this.

Also, it's true that with pydantic 2.3.0 there's no `Internal Server Error` but there are other errors when trying to parse `FlowResponse` to JSON.
<img width="1519" alt="Screenshot 2023-10-20 at 11 03 55" src="https://github.com/tiangolo/fastapi/assets/64859146/33a1c214-0e72-469e-8f10-206e71e9688a">

**Comment by atompie** [comment]
Hi @Ale-Cas 

The example is a simplification. I use it in post/put methods with some more complex objects. I noticed that adding any arbitrary type makes fastAPI fail. I added get as it is easier to replicate the error not particular use case with get.

**Comment by atompie** [comment]
@Ale-Cas FYI I also posted the issue on pydantic and it seems that this is fastAPI issue.
